clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1160,15 +1160,35 @@ static target getAccessTarget(QualType FieldTy,
 }
 
 bool SemaSYCL::isFreeFunction(const FunctionDecl *FD) {
-  for (auto *IRAttr : FD->specific_attrs<SYCLAddIRAttributesFunctionAttr>()) {
-    SmallVector<std::pair<std::string, std::string>, 4> NameValuePairs =
-        IRAttr->getAttributeNameValuePairs(getASTContext());
-    for (const auto &NameValuePair : NameValuePairs) {
-      if (NameValuePair.first == "sycl-nd-range-kernel" ||
-          NameValuePair.first == "sycl-single-task-kernel") {
-        return true;
+  llvm::SmallVector<clang::FunctionDecl *, 4> Redecls;
+  for (clang::FunctionDecl *Redecl : FD->redecls()) {
+    Redecls.push_back(Redecl);
+  }
+  bool FirstDecl = true;
+  clang::SourceLocation Loc = FD->getLocation();
+  while (!Redecls.empty()) {
+    auto *Redecl = Redecls.back();
+    Redecls.pop_back();
+    if (FirstDecl)
+      Loc = Redecl->getLocation(); // Save the location of the first decl to use
+                                   // in diagnostics.
+    for (auto *IRAttr :
+         Redecl->specific_attrs<SYCLAddIRAttributesFunctionAttr>()) {
+      SmallVector<std::pair<std::string, std::string>, 4> NameValuePairs =
+          IRAttr->getAttributeNameValuePairs(getASTContext());
+      for (const auto &NameValuePair : NameValuePairs) {
+        if (NameValuePair.first == "sycl-nd-range-kernel" ||
+            NameValuePair.first == "sycl-single-task-kernel") {
+          if (FirstDecl)
+            return true;
+          else {
+            Diag(Loc, diag::err_free_function_first_occurrence_missing_attr);
```

Comment:
to make it closer to overall clang style, I would instead emit error at not the first declaration with the attribute with a note saying "first declared here" i.e. for example similar to https://godbolt.org/z/MaoaovMdW